src/daft-dsl/src/functions/mod.rs

```diff
@@ -123,40 +123,55 @@ pub fn function_semantic_id(func: &FunctionExpr, inputs: &[ExprRef], schema: &Sc
     FieldID::new(format!("Function_{func:?}({inputs})"))
 }
 
+/// FunctionRegistry is a lookup structure for scalar functions.
 #[derive(Default)]
 pub struct FunctionRegistry {
     // Todo: Use the Bindings object instead, so we can get aliases and case handling.
-    map: HashMap<String, Arc<dyn ScalarUDF>>,
+    map: HashMap<String, Arc<dyn ScalarFunctionFactory>>,
 }
+
+/// FunctionModule is a mechanism to group and register scalar functions.
 pub trait FunctionModule {
     /// Register this module to the given [SQLFunctions] table.
     fn register(_parent: &mut FunctionRegistry);
 }
 
 impl FunctionRegistry {
+    /// Creates an empty FunctionRegistry.
     pub fn new() -> Self {
         Self {
             map: HashMap::new(),
         }
     }
+
+    /// Registers all functions defined in the FunctionModule to this registry.
     pub fn register<Mod: FunctionModule>(&mut self) {
         Mod::register(self);
     }
 
-    pub fn add_fn<UDF: ScalarUDF + 'static>(&mut self, func: UDF) {
-        let func = Arc::new(func);
-        // todo: use bindings instead of hashmap so we don't need duplicate entries.
-        for alias in func.aliases() {
-            self.map.insert((*alias).to_string(), func.clone());
+    /// Registers a scalar function factory.
+    pub fn add_function(&mut self, function: impl ScalarFunctionFactory + 'static) {
```

Comment:
Using `impl ScalarFunctionFactory` rather than a generic because I don't want monomorphization. This is one-time initialization code, and we store as a dyn anyways. The existing `add_fn` will monomorphize for _every_ function we define which isn't favorable since we'll end up generating a unique `add_fn` for literally every function.